README.md
> :warning: You are looking at the README for the master branch of this gem.
> See the latest [released version (1.0.1)](https://github.com/influxdata/influxdb-rails/tree/v1.0.1#readme)
> instead.

# influxdb-rails

[![Gem Version](https://badge.fury.io/rb/influxdb-rails.svg)](https://badge.fury.io/rb/influxdb-rails)
[![Build Status](https://github.com/influxdata/influxdb-rails/actions/workflows/spec.yml/badge.svg)](https://github.com/influxdata/influxdb-rails/actions)

Automatically instrument your Ruby on Rails applications and write the metrics directly into
[InfluxDB](https://www.influxdata.com/).

## Table of contents

- [Installation](#installation)
- [Usage](#installation)
- [Configuration](#configuration)
- [Demo](#demo)
- [FAQ](#frequently-asked-questions)
- [Contributing](#contributing)

## Installation

Add the gem to your `Gemfile`:

```console
echo 'gem "influxdb-rails"' >>Gemfile
bundle install
```

To get things set up, create an initializer:

```console
bundle exec rails generate influxdb
```

This creates a file `config/initializers/influxdb_rails.rb`, which allows
configuration of this gem.

## Usage

Out of the box, you'll automatically get reporting for the Ruby on Rails components mentioned
below.

### Action Controller

Reported ActiveSupport instrumentation hooks:

- [start\_processing.action\_controller](https://guides.rubyonrails.org/active_support_instrumentation.html#start-processing-action-controller)
- [process\_action.action\_controller](https://guides.rubyonrails.org/active_support_instrumentation.html#process-action-action-controller)

Reported values:

```ruby
  controller: 48.467,
  view: 46.848,
  db: 0.157,
  started: 1465839830100400200,
  request_id: "d5bf620b-3494-425b-b7e1-4953597ea744"
```

Reported tags:

```ruby
{
  hook:        "process_action",
  server:      Socket.gethostname,
  app_name:    configuration.application_name,
  method:      "PostsController#index",
  http_method: "GET",
  format:      "html",
  status:      ["200", ""],
  exception:   "ArgumentError"
}
```

*Note*: If an exception happens during that particular action the `status` will be blank and the tag `exception` will contain the name of the exception class. The status is blank because we can't know how you handle the exception outside the action.

### Action View

Reported ActiveSupport instrumentation hooks:

- [render\_template.action\_view](https://guides.rubyonrails.org/active_support_instrumentation.html#render-template-action-view)
- [render\_partial.action\_view](https://guides.rubyonrails.org/active_support_instrumentation.html#render-partial-action-view)
- [render\_collection.action\_view](https://guides.rubyonrails.org/active_support_instrumentation.html#render-collection-action-view)

Reported values:

```ruby
  value: 48.467,
  count: 3,
  cache_hits: 0,
  request_id: "d5bf620b-3494-425b-b7e1-4953597ea744"
```

Reported tags:

```ruby
  hook:       ["render_template", "render_partial", "render_collection"],
  server:     Socket.gethostname,
  app_name:   configuration.application_name,
  location:   "PostsController#index",
  filename:   "/some/file/action.html"
```

### Active Record

Reported ActiveSupport instrumentation hooks:

- [sql.active\_record](https://guides.rubyonrails.org/active_support_instrumentation.html#sql-active-record)
- [instantiation.active\_record](https://guides.rubyonrails.org/active_support_instrumentation.html#instantiation-active-record)

Reported SQL values:

```ruby
  sql: "SELECT \"posts\".* FROM \"posts\"",
  request_id: "d5bf620b-3494-425b-b7e1-4953597ea744"
```

Reported SQL tags:

```ruby
  hook:       "sql",
  server:     Socket.gethostname,
  app_name:   configuration.application_name,
  location:   "PostsController#index",
  operation:  "SELECT",
  class_name: "POST",
  name:       "Post Load"
```

Reported instantiation values:

```ruby
  record_count: 1,
  request_id:   "d5bf620b-3494-425b-b7e1-4953597ea744"
  value:        7.689
```

Reported instantiation tags:

```ruby
  hook:       "instantiation",
  server:     Socket.gethostname,
  app_name:   configuration.application_name,
  location:   "PostsController#index",
  class_name: "POST"
```

### Active Job

Reported ActiveSupport instrumentation hooks:

- [enqueue.active\_job](https://guides.rubyonrails.org/active_support_instrumentation.html#enqueue-active-job)
- [perform.active\_job](https://guides.rubyonrails.org/active_support_instrumentation.html#perform-active-job)

Reported values:

```ruby
  value: 89.467
```

Reported tags:

```ruby
  hook:       ["enqueue", "perform"],
  state:      ["queued", "succeeded", "failed"],
  job:        "SomeJobClassName",
  queue:      "queue_name"
```

*Note*: Only the measurements with the hook `perform` report a duration in the value.
The enqueue hook is a counter and always reports a value of `1`.

### Action Mailer

Reported ActiveSupport instrumentation hooks:

- [deliver.action\_mailer](https://guides.rubyonrails.org/active_support_instrumentation.html#deliver-action-mailer)

Reported values:

```ruby
  value: 1
```

Reported tags:

```ruby
  hook:               "deliver",
  mailer:             "SomeMailerClassName"
```

*Note*: The hook is just a counter and always report a value of `1`.

## Configuration

The only setting you actually need to configure is the name of the database
within the InfluxDB server instance (don't forget to create this database!).

```ruby
InfluxDB::Rails.configure do |config|
  config.client.database = "rails"
end
```

You'll find all of the configuration settings in the initializer file.

### Custom Tags

You can modify the tags sent to InfluxDB by defining a middleware, which
receives the current tag set as argument and returns a hash in the same
form. The middleware can be any object, as long it responds to `#call`
(like a `Proc`):

```ruby
InfluxDB::Rails.configure do |config|
  config.tags_middleware = lambda do |tags|
    tags.merge(env: Rails.env)
  end
end
```

The `tags` argument is a Hash (mapping Symbol keys to String values). The
actual keys and values depend on the series name (`tags[:series]`, see
next section).

If you want to add dynamically tags or fields *per request*, you can access
`InfluxDB::Rails.current` to do so. For instance, you could add the current
user as tag or redis query time to every data point:

```ruby
class ApplicationController
  before_action :set_influx_data

  def set_influx_data
    InfluxDB::Rails.current.tags = { user: current_user.id }
    InfluxDB::Rails.current.values = { redis_value: redis_value }
  end
end
```

### Block Instrumentation
If you want to add custom instrumentation, you can wrap any code into a block instrumentation

```ruby
InfluxDB::Rails.instrument "expensive_operation", tags: { }, values: { } do
  expensive_operation
end
```

Reported tags:

```ruby
  hook:       "block_instrumentation",
  server:     Socket.gethostname,
  app_name:   configuration.application_name,
  location:   "PostsController#index",
  name:       "expensive_operation"
```

Reported values:
```ruby
  value: 100 # execution time of the block in ms
```

You can also overwrite the `value`

```ruby
InfluxDB::Rails.instrument "user_count", values: { value: 1 } do
  User.create(name: 'mickey', surname: 'mouse')
end
```

or call it even without a block

```ruby
InfluxDB::Rails.instrument "temperature", values: { value: 25 }
```

### Custom client configuration

The settings named `config.client.*` are used to construct an `InfluxDB::Client`
instance, which is used internally to transmit the reporting data points
to your InfluxDB server. You can access this client as well, and perform
arbitrary operations on your data:

```ruby
InfluxDB::Rails.client.write_point "events",
  tags:   { url: "/foo", user_id: current_user.id, location: InfluxDB::Rails.current.location },
  values: { value: 0 }
```

If you do that, it might be useful to add the current context to these custom
data points which can get accessed with `InfluxDB::Rails.current.location`.

See [influxdb-ruby](http://github.com/influxdata/influxdb-ruby) for a
full list of configuration options and detailed usage.

### Disabling hooks

If you are not interested in certain reports you can disable them in the configuration.

```ruby
InfluxDB::Rails.configure do |config|
  config.ignored_hooks = ['sql.active_record', 'render_template.action_view']
end
```

## Demo
Want to see this in action? Check out our [sample dashboard](https://github.com/influxdata/influxdb-rails/tree/master/sample-dashboard).

## Frequently Asked Questions

### I'm seeing far less requests recorded in InfluxDB than my logs suggest.

By default, this gem writes data points with *millisecond time precision*
to the InfluxDB server. If you have more than 1000 requests/second (and/or
multiple parallel requests), **only the last** data point (within the
same tag set) is stored. See [InfluxDB server docs][duplicate-points] for
further details.

To work around this limitation, set the `config.client.time_precision`
to one of `"us"` (microseconds, 1·10<sup>-6</sup>s) or `"ns"` (nanoseconds,
1·10<sup>-9</sup>s).

Please note: This will only ever reduce the likelihood of data points
overwriting each other, but not eliminate it completely.

[duplicate-points]: https://docs.influxdata.com/influxdb/v1.4/troubleshooting/frequently-asked-questions/#how-does-influxdb-handle-duplicate-points


### How does the measurement influence the response time?

This gem subscribes to various `ActiveSupport::Notifications` hooks.
(cf. [guide][arn-guide] · [docs][arn-docs] · [impl][arn-impl]). The
controller notifications are run *after* a controller action has finished,
and should not impact the response time.

Other notification hooks (rendering and SQL queries) run *inline* in the
request processing. The amount of overhead introduced should be negligible,
though. However reporting of SQL queries relies on Ruby string parsing which
might cause performance issues on traffic intensive applications. Disable it in
the configuration if you are not willing to tolerate this.

By default, this gem performs writes to InfluxDB asynchronously. A single
hook usually only performs some time delta calculations, and then enqueues
the data point into a worker queue (which is processed by a background
thread).

If you, however, use a synchronous client (`config.client.async = false`),
the data points are immediately sent to the InfluxDB server. Depending on
the network link, this might cause the HTTP thread to block a lot longer.

[arn-guide]: http://guides.rubyonrails.org/v5.1/active_support_instrumentation.html#process-action-action-controller
[arn-docs]: http://api.rubyonrails.org/v5.1/classes/ActiveSupport/Notifications.html
[arn-impl]: https://github.com/rails/rails/blob/5-1-stable/actionpack/lib/action_controller/metal/instrumentation.rb#L30-L38

### How does this gem handle an unreachable InfluxDB server?

By default, the InfluxDB client will retry indefinitely, until a write
succeeds (see [client docs][] for details). This has two important
implications, depending on the value of `config.client.async`:

- if the client runs asynchronously (i.e. in a separate thread), the queue
  might fill up with hundreds of megabytes of data points
- if the client runs synchronously (i.e. inline in the request/response
  cycle), it might block all available request threads

In both cases, your application server might become unresponsive and needs
to be restarted.

If you setup a maximum retry value (`Integer === config.client.retry`),
the client will try up to that amount of times to send the data to the server
and (on final error) log an error and discard the values.

[client docs]: https://github.com/influxdata/influxdb-ruby#retry

### What happens, when the InfluxDB client or this gem throws an exception? Will the user see 500 errors?

No. The controller instrumentation is wrapped in a `rescue StandardError`
clause, i.e. this gem will only write the error to the `client.logger`
(`Rails.logger` by default) and not disturb the user experience.

### What happens with unwritten points, when the application restarts?

The data points are simply discarded.

## Contributing

- Fork this repository on GitHub
- Make your changes
  - Add tests.
  - Add an entry in the `CHANGELOG.md` in the "unreleased" section on top.
- Run the tests:
  - Either run them manually:

    ```console
    rake test:all
    ```

  - or wait for [our CI](https://github.com/influxdata/influxdb-rails/actions) to pick up your changes, *after*
    you made a pull request.
- Send a pull request.
- If your changes are looking good, we'll merge them.

### Testing Tasks

```console
rake            # unit tests + Rubocop linting
rake spec       # only unit tests
rake rubocop    # only Rubocop linter
rake test:all   # integration tests with various Rails version
```


CHANGELOG.md
# Changelog

For the full commit log, [see here](https://github.com/influxdata/influxdb-rails/commits/one-stable).

## v2.0.0.beta1, not released yet
- Drop support for Ruby < 3.0, Rails < 7.0

## v1.0.2, released 2021-04-13

Fix passing booleans to custom tags.

## v1.0.1, released 2021-03-23

The final release, no code changes.

## v1.0.1.beta3, released 2020-10-16

- Drop `perform_start.active_job` subscriber, it's rather useless on dashboards.

## v1.0.1.beta2, released 2020-09-14

- Implement [`deliver.action_mailer`](https://guides.rubyonrails.org/active_support_instrumentation.html#deliver-action-mailer) subscriber
- Add missing Active Job documentation
- Drop support for Ruby 2.4
- Drop support for Rails < 5.2

## v1.0.1.beta1, released 2020-08-21

- Drop support for Ruby 2.3
- Drop support for Rails 4.x
- Add `auth_method` to client configuration (#96, @anlek)
- Drop undocumented `instrumentation_enabled` setting, use
  `ignored_environments` do disable instrumentation
- Simplified spec with a PORO test client
- Implement `instantiation.active_record` subscriber (https://guides.rubyonrails.org/active_support_instrumentation.html#instantiation-active-record)
- Implement `enqueue.active_job` subscriber (https://guides.rubyonrails.org/active_support_instrumentation.html#enqueue-active-job)
- Implement `perform_start.active_job` subscriber (https://guides.rubyonrails.org/active_support_instrumentation.html#perform-start-active-job)
- Implement `perform.active_job` subscriber (https://guides.rubyonrails.org/active_support_instrumentation.html#perform-active-job)
- Implement block instrumentation `InfluxDB::Rails.instrument do; 1 + 1; end`
- Record unhandled exceptions as tags for process_action.action_controller

## v1.0.0, released 2019-10-23

The Final release, no code changes.

## v1.0.0.beta5, unreleased

- Silently eat all dropped configuration options and do not crash
- Add per action view to the sample dashboard

## v1.0.0.beta4

- Introduces a Sample Grafana Dashboard + docker-compose demo (#75/#79, @hennevogel)
- Redesign Measurement Output (#66, @dmke, @ChrisBr, @hennevogel)
  - Switching from emitting eight different measurements to one called `rails`
    to support easier aggregation across data, to simplify configuration and to
    stay closer to the InfluxDB/Grafana nomenclature.
- Introduced configuration option `ignored_hooks` to disable specifc hooks
- Enable SQL subscriber by default
  - Set a default location (:raw) for SQL subscriber
- Add dynamic values (#65, @ChrisBr)
- Remove empty tags (#64, @ChrisBr)

## v1.0.0.beta3, released 2019-01-07

- Add dynamic tags (#62, @ChrisBr)
- Reduce cardinality of resulting InfluxDB measurement by moving
  some tags to values (#63, @ChrisBr)
- Remove SCHEMA queries from SQL instrumentation (#61, @ChrisBr)

## v1.0.0.beta2, released 2018-12-07

- Added `tags_middleware` config option (#47, @Kukunin)
- Removed path tag from metrics (introduced with #50), because it
  potentially produces "exceed tag value limit" (#54, @ChrisBr)
- Added render instrumentation (#53, @ChrisBr)
- Added SQL instrumentation (#55, @ChrisBr)

## v1.0.0.beta1, released 2018-11-22

- Added app name to the measurement's tag sets (#44, @stefanhorning)
- Added config parameters for additional series:
  - `InfluxDB::Rails::Configuration#series_name_for_instrumentation`
  - `InfluxDB::Rails::Configuration#series_name_for_exceptions`
- Added method, status and format tags to metrics (#50, @ChrisBr)

### Breaking changes

- Support for Ruby <= 2.2.x has been removed
- Support for Rails <= 4.1.x has been removed
- Removed previously deprecated methods:
  - `InfluxDB::Rails::Configuration#reraise_global_exceptions`
  - `InfluxDB::Rails::Configuration#database_name`
  - `InfluxDB::Rails::Configuration#application_id`

## v0.4.3, released 2017-12-12

- Added `time_precision` config option (#42, @kkentzo)

## v0.4.2, released 2017-11-28

- Added `open_timeout`, `read_timeout`, and `max_delay` config options
  (#41, @emaxi)
- Deprecate unused method (`#reraise_global_exceptions` in
  `InfluxDB::Rails::Configuration`, #37, @vassilevsky)

## v0.4.1, released 2017-10-23

- Bump `influx` version dependency (#40, @rockclimber73)

## v0.4.0, released 2017-08-19

- Drop support for Rails 3, Ruby < 2.2
- Sync version with `influxdb` gem

## v0.1.12, released 2017-06-06

- Added Rails 5.1 compatibility (#31, @djgould).
- Added `retry` config option (#17, @scambra and #18, @bcantin).

## v0.1.11, released 2016-11-24

- Bumped `influxdb` Rubygem dependency (#28, #32, @agx).

**Note:** The real changelog starts here. Previous entries are reconstructed
from the commit history by correlating release version with Git tags, which
may or may not reflect what was really released.

## v0.1.10, released 2014-10-08

- Lazy loading of `InfluxDB::Client` (#15, @chingo13).
- Various test fixes.

## v0.1.9, released 2014-06-18

- v0.1.8 was yanked
- Initializer now allows multiple hosts.

## v0.1.7, released 2014-04-09

- Added logger.

## v0.1.6, released 2014-04-08

- No changes (?)

## v0.1.5, released 2014-04-02

- No changes (?)

## v0.1.4, released 2014-03-26

- Added docs
- Made `async` option default to `true`.

## v0.1.3, released 2014-02-11

- Set series name defaults.
- Fixed a configuration bug.

## v0.1.2, released 2014-02-11

- Graceful handling of authentication errors.
- Fixes in documentation.

## v0.1.1, released 2014-02-06

- Refactoring of `ActiveSupport::Notification` handling.
- Added more initializer options.

## v0.1.0, released 2014-02-06

- Larger refactoring.


lib/influxdb/railtie.rb
require 'influxdb'
require 'rails'

module InfluxDB
  class Railtie < ::Rails::Railtie
    rake_tasks do
      namespace :influxdb do
        task :test => :environment do
          if InfluxDB.configuration.api_key.nil?
            puts "Hey, you need to define an API key first. Run `rails g influxdb <api-key>` if you didn't already."
            exit
          end

          InfluxDB.configure do |config|
            config.debug = true
            config.ignored_environments = []
          end

          data = [{:n => "tests", :p => [{:v => 1}]}]
          if InfluxDB.api.post(data)
            puts "Test data sent successfully!"
          else
            puts "Test failed! Check your network connection and try again."
          end
        end

        task :diagnose => :environment do
          if InfluxDB.configuration.api_key.nil?
            puts "Hey, you need to define an API key first. Run `rails g influxdb <api-key>` if you didn't already."
            exit
          end

          InfluxDB.configure do |config|
            config.ignored_environments = []
          end

          class ::InfluxDBSampleException < Exception; end;

          require ::Rails.root.join("app/controllers/application_controller.rb")

          puts "Setting up ApplicationController.."
          class ::ApplicationController
            prepend_before_filter :raise_sample_exception

            def raise_sample_exception
              raise ::InfluxDBSampleException.new("If you see this, InfluxDB is working.")
            end

            def influxdb_dummy_action; end
          end

          ::Rails.application.routes_reloader.execute_if_updated
          ::Rails.application.routes.draw do
            match "influxdb_test" => 'application#influxdb_dummy_action'
          end

          puts "Generating sample request.."
          env = ::Rack::MockRequest.env_for("/influxdb_test")

          puts "Attempting to raise exception via HTTP.."
          response = ::Rails.application.call(env)

          10.times do
            sleep 1
            break unless InfluxDB.api.last_response.nil?
          end

          if response.try(:first) == 500
            if InfluxDB.api.last_response.nil?
              puts "Uh oh. Your app threw an exception, but we didn't get a response. Check your network connection and try again."
            elsif InfluxDB.api.last_response.code == "201"
              puts "Done. Check your email or http://influxdb.com for the exception notice."
            else
              puts "That didn't work. The InfluxDB API said: #{InfluxDB.api.last_response.body}"
            end
          else
            puts "Request failed: #{response}"

            env["HTTPS"] = "on"
            puts "Attempting to raise exception via HTTPS.."
            response = ::Rails.application.call(env)

            if response.try(:first) == 500
              if InfluxDB.api.last_response.nil?
                puts "Uh oh. Your app threw an exception, but we didn't get a response. Check your network connection and try again."
              elsif InfluxDB.api.last_response.code == "201"
                puts "Done. Check your email or http://influxdb.com for the exception notice."
              else
                puts "That didn't work. The InfluxDB API said: #{InfluxDB.api.last_response.body}"
              end
            else
              puts "Request failed: #{response}"
              puts "We didn't get the exception we were expecting. Contact support@influxdb.com and send them all of this output."
            end
          end
        end
      end
    end

    initializer "influxdb.insert_rack_middleware" do |app|
      app.config.middleware.insert 0, InfluxDB::Rack
    end

    config.after_initialize do
      InfluxDB.configure(true) do |config|
        config.logger                ||= ::Rails.logger
        config.environment           ||= ::Rails.env
        config.application_root      ||= ::Rails.root
        config.application_name      ||= ::Rails.application.class.parent_name
        config.framework               = "Rails"
        config.framework_version       = ::Rails.version
      end

      ActiveSupport.on_load(:action_controller) do
        require 'influxdb/rails/air_traffic_controller'
        include InfluxDB::Rails::AirTrafficController
        require 'influxdb/rails/instrumentation'
        include InfluxDB::Rails::Instrumentation
      end

      if defined?(::ActionDispatch::DebugExceptions)
        require 'influxdb/rails/middleware/hijack_render_exception'
        ::ActionDispatch::DebugExceptions.send(:include, InfluxDB::Rails::Middleware::HijackRenderException)
      elsif defined?(::ActionDispatch::ShowExceptions)
        require 'influxdb/rails/middleware/hijack_render_exception'
        ::ActionDispatch::ShowExceptions.send(:include, InfluxDB::Rails::Middleware::HijackRenderException)
      end

      if defined?(ActiveSupport::Notifications)
        ActiveSupport::Notifications.subscribe "process_action.action_controller" do |name, start, finish, id, payload|
          if InfluxDB.configuration.instrumentation_enabled  && ! InfluxDB.configuration.ignore_current_environment?
            timestamp = finish.utc.to_i
            controller_runtime = ((finish - start)*1000).ceil
            view_runtime = (payload[:view_runtime] || 0).ceil
            db_runtime = (payload[:db_runtime] || 0).ceil
            controller_name = payload[:controller]
            action_name = payload[:action]

            dimensions = {:method => "#{controller_name}##{action_name}", :server => Socket.gethostname}
            InfluxDB.aggregate "controllers", :value => controller_runtime, :dimensions => dimensions
            InfluxDB.aggregate "views", :value => view_runtime, :dimensions => dimensions
            InfluxDB.aggregate "db", :value => db_runtime, :dimensions => dimensions
          end
        end
      end
    end
  end
end
